bookkeeper-server/src/test/java/org/apache/bookkeeper/client/impl/LedgerEntriesImplTest.java

```diff
@@ -0,0 +1,130 @@
+    private final List<LedgerEntry> entryList = Lists.newArrayList();
+
+    // content for each entry
+    private final long ledgerId = 1234L;
+    private final long entryId = 5678L;
+    private final long length = 9876L;
+    private final byte[] dataBytes = "test-ledger-entry-impl".getBytes(UTF_8);
+
+    public LedgerEntriesImplTest () {
+        for(int i = 0; i < entryNumber; i++) {
+            entryList.add(LedgerEntryImpl.create(ledgerId + i,
+                entryId + i,
+                length + i,
+                Unpooled.wrappedBuffer(dataBytes)));
+        }
+
+        ledgerEntriesImpl = LedgerEntriesImpl.create(entryList);
+    }
+
+    @After
+    public void tearDown() {
+        ledgerEntriesImpl.close();
+
+        try {
+            ledgerEntriesImpl.getEntry(entryId);
+            fail("should fail getEntry after close");
+        } catch (NullPointerException e) {
+
+        }
+
+        try {
+            ledgerEntriesImpl.iterator();
+            fail("should fail iterator after close");
+        } catch (NullPointerException e) {
+
+        }
+
+        try {
+            ledgerEntriesImpl.retainIterator();
+            fail("should fail retainIterator after close");
+        } catch (NullPointerException e) {
+
+        }
+    }
+
+    @Test
+    public void testGetEntry() {
+        for(int i = 0; i < entryNumber; i ++) {
+            LedgerEntry entry = ledgerEntriesImpl.getEntry(entryId + i);
+            assertEquals(entryList.get(i).getLedgerId(),  entry.getLedgerId());
+            assertEquals(entryList.get(i).getEntryId(),  entry.getEntryId());
+            assertEquals(entryList.get(i).getLength(),  entry.getLength());
+
+            ByteBuf buf = entry.getEntryBuffer();
+            byte[]  content = new byte[buf.readableBytes()];
+            buf.readBytes(content);
+            assertArrayEquals(dataBytes, content);
+
+            assertEquals(1, entry.getEntryBuffer().refCnt());
+        }
+
+        try {
+            LedgerEntry entry = ledgerEntriesImpl.getEntry(entryId - 1);
+            fail("Should get IndexOutOfBoundsException");
+        } catch (IndexOutOfBoundsException e) {
+
+        }
+
+        try {
+            LedgerEntry entry = ledgerEntriesImpl.getEntry(entryId + entryNumber);
+            fail("Should get IndexOutOfBoundsException");
+        } catch (IndexOutOfBoundsException e) {
+
+        }
+    }
+
+    @Test
+    public void testIterator() {
+        Iterator<LedgerEntry> entryIterator = ledgerEntriesImpl.iterator();
+        entryIterator.forEachRemaining(ledgerEntry -> assertEquals(1, ledgerEntry.getEntryBuffer().refCnt()));
+    }
+
+    @Test
+    public void testRetainIterator() {
+        Iterator<LedgerEntry> entryIterator = ledgerEntriesImpl.retainIterator();
+        entryIterator.forEachRemaining(ledgerEntry -> assertEquals(2, ledgerEntry.getEntryBuffer().refCnt()));
+        entryIterator.forEachRemaining(ledgerEntry -> ledgerEntry.getEntryBuffer().release());
```

Comment:
I think this will not be tested. `forEachRemaining` will consume the iterator.

you might need change this to 

```
entryIterator.forEachRemaining(ledgerEntry -> {
    assertEquals(2, ledgerEntry.getEntryBuffer().refCnt());
    ledgerEntry.getEntryBuffer().release();
    assertEquals(1, ledgerEntry.getEntryBuffer().refCnt());
})
```

Reply:
Thanks. will change this